man/vzeventd.8
.TH vzeventd 8 "28 Jun 2011" "OpenVZ" "Containers"
.SH NAME
vzeventd \- the OpenVZ events daemon.
.SH SYNOPSIS
.SY vzeventd
.OP \-v
.OP \-d
.YS
.SY vzeventd
.B \-h
.YS
.SH DESCRIPTION
This daemon takes care of events sent by the OpenVZ kernel
(via a netlink socket) and performs required actions associated with
those events, by running specific scripts. Every event received contains
a container ID, which is passed to the script as \fBVEID\fR environment
variable.
.P
Current list of known events and associated \fBvzeventd\fR actions are:
.TP
.B start
Ignore.
.TP
.B stop
Run /usr/lib/vzctl/scripts/vzevent-stop. This script takes care of removing
ARP and routing records for the given CT from CT0.
.TP
.B mount
Ignore.
.TP
.B umount
Ignore.
.TP
.B reboot
Run /usr/lib/vzctl/scripts/vzevent-reboot. This script takes care of rebooting
a given CT.
.SH OPTIONS
.TP
.B \-v
Increase verbosity (can be used multiple times).
.TP
.B \-d
Debug mode (do not daemonize, run in foreground).
.TP
.B -h
Display help and exit.
.SH EXIT STATUS
Returns 0 upon success.
.SH LICENSE
Copyright (C) 2010-2011, Parallels, Inc. Licensed under GNU GPL.
.SH AUTHOR
.PP
This manual page was initially written by Thorsten Schifferdecker
<tsd@debian.systs.org>
for the \fBDebian GNU/Linux\fP system (but may be used by others).


ChangeLog
Since after 3.0.12 release, this file is obsoleted. Complete change log
can be obtained from OpenVZ git repository.

	http://git.openvz.org/?p=vzctl;a=shortlog
	http://git.openvz.org/?p=vzctl;a=log

Below is a change log for old vzctl releases.


   3.0.12: Thu Oct 5 2006 <igor@openvz.org>
	* vzctl start: added --wait cmdline option
	* added vznetcfg script for veth setup
	* fixed error message in case kernel compiled without USER_RESOURCES
	* added '--applyconfig_map name' vzctl set option
	* changed friendly name data structure
	* fixed vzmigrate to support VE name feature (bug #239)
   	* fixed annoying arp warning (bug #189)
	* added bash completion script (bug #212)
	* added check to vzmigrate if OpenVZ is run on another side (bug #163)
	* removed debian directory from the tarball (bugs #197, #194)
	* removed LICVIEW from /etc/init.d/vz (bug #237)
	* fixed misprints and /dev/nul(l) usage in dist scripts (bug #228)
	* fixed vzmigrate to work with different awk flavors (bug #231)
	* years in copyright fixed (bug #232)
	* increased some UBCs in ve-light.conf-sample (bug #234)
	* fixed some minor memory leaks while doing realloc()
	* vzctl destroy: remove VE names on destroy (bug #236)
	* added --name option description to vzctl(8) man page (bug #232)
	* merged libvzctl-simfs and libvzctl into a single libvzctl
	* vzlist: replaced sprintf() with snprintf() to avoid buffer overflows
	* vzctl enter: forward error messages from child process
	* vzctl enter: fix off-by-one error which resulted in
	  uninitialized PATH
	* arpsend: fixed segfault when parsing long options
	* vzlist: fixed cast from pointer to integer of different size warnings
	* vzctl enter: exit with correct error code on enter fail (bug #245)
	* added udev rules for /dev/vzctl (bug #204)
	* added syscall numbers for powerpc
	* /etc/init.d/vz: check for /dev/vzctl only after modprobe (bug #204)
	* fixed ppc64 compilation
	* removed distro-specific stuff from 'make install' (bug #254)
	* added 'make install-gentoo' and Gentoo init.d/vz script (bug #262)
	* fixed arpsend options parsing for some compilers (bug #261)
	* man pages cleanup (VPS->VE, etc.)
	* changed all utilities and man pages to refer to VE not VPS
	* vzcpucheck: give error if VE0 CPU units not set (bug #249)

   3.0.11: Mon Aug  7 2006 <igor@openvz.org>
	* fixed using VE_ROUTE_SRC_DEV variable
	* added VPS name support
	* added change number of CPUs available in the VPS functionality.
	* fixes in CPT error processing
	* added upport of xt_mac iptables module
	* fixed awk expression to be compatible with mawk
   3.0.10: Wed Jun  7 2006 <igor@openvz.org>
	* added virtual ethernet device support (--veth_add, --veth_del)
	* added  /proc/meminfo virtualization support (--meminfo mode:value)
	* added VE0 UB setup functionality (--reset-ub)
   3.0.9: Wed May 24 2006 <igor@openvz.org>
	vzctl:
	* switched to new versioning scheme.
	* added -z option to /sbin/init
		(http://bugzilla.openvz.org/show_bug.cgi?id=164)
	* reset standart file desc [0-2] to /dev/null in case they are closed.
	* added possibility to set cpulimit in %
	* added suffixes G,M,K support for --diskspace option
	* set SIGPIPE handler to SIG_IGN before fist write to stdout/err
	* added Debian packaging support
	* changed directory structure.
		/etc/sysconfig/vz-scripts -> /etc/vz/conf
		/etc/sysconfig/vz-scripts/dists -> /etc/vz/dists
		/etc/sysconfig/vz -> /etc/vz/vz.conf
	* better distribution detection in vz initscript.
	* added Ubuntu distribution configuration file as copy of debian.
		(http://bugzilla.openvz.org/show_bug.cgi?id=119)
	* Increased kmemsize, privmpages, numfiles for x86_64
		ve-vps.basic.conf-sample.
		(http://bugzilla.openvz.org/show_bug.cgi?id=69)
	* changed -p -> --preserve=mode,ownership for 'cp' workaround in
		case /proc is not mounted.
	* internal modifications
		changed dynamic UB structure to static one.
		moved VPS configure functionality to vps_configure.c
		moved destroy functionaliny to destroy.c
	vzmigrate:
	* move config as veid.migrated to allow backward migration
		(http://bugzilla.openvz.org/show_bug.cgi?id=152) 
	* added coping of action scripts.
		(http://bugzilla.openvz.org/show_bug.cgi?id=157)
	* fixed --keep-dst option processing. Option behavior was inverted.
   3.0.0-8: Mon Apr 24 2006 <igor@openvz.org>
	* UBC parameters > LONG_MAX in VPS conf truncate to LONG_MAX
	* added installation of vzmigrate.8 man page
	* use 'uname -m' for arch selection	
   3.0.0-6: Tue Apr 18 2006 <kir@openvz.org>
	* man page fixes
   3.0.0-5: Fri Apr  7 2006 <igor@sw.ru>
	* added checkpointing support
	* added kernel headers for standalone compilation 
	* chaged license from QPL to GPL
	* set personality PER_LINUX32 on x86_64 arch for i386 VPSes
	* fixed --iptables store (do not store dependent modules)
	* Do skip unknown iptables modules in config
 	* set arg[0] = "-bash" on enter
   3.0.0-4: Thu Mar 16 2006 <kir@openvz.org>
	* cron scripts do nothing if VZ is not loaded (bug #107)
	* cosmetic vzctl.8 man page fixes
	* reverted bug #110 changes, see http://bugzilla.openvz.org/111
	* added ipt_REDIRECT support, thanks to jstubbs-at-gentoo.org
   3.0.0-3: Thu Mar  2 2006 <igor@sw.ru>
	* put_param3() can incorrectly set $name=( "" ) (bug #111).
		Affects gentoo VPSs. Thanks to jstubbs-at-gentoo.org
	* fixed compilation against old kernels (<= 022stab070)
	* description of set parameter suffixes (K, M, G, P) added
		to vzctl.8 man page (bug #105)
	* fixed gentoo-add_ip.sh script (bug #110)
   3.0.0-2: Thu Feb 23 2006 <igor@sw.ru>
	* fixed VPS start on old kernel (witthout feature_mask support)	
   3.0.0-1: Wed Feb 22 2006 <igor@sw.ru>
	* enable sysfs for SuSE like distributions
	* added possibility to setup UBC parameters in the form
		xxx[GMKPB]:yyy[GMKPB]
   2.7.0-27: Tue Feb  7 2006 <igor@sw.ru>
	* fixed VE_CONFFILE variable passed to vps.mount script
	* fixed venet0 initialization in vz initscript (bug #99)
   2.7.0-26: Thu Feb  2 2006 <igor@sw.ru>

src/vzeventd.c
/*
 *  Copyright (C) 2010-2011, Parallels, Inc. All rights reserved.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#include <sys/types.h>
#include <sys/socket.h>
#include <linux/types.h>
#include <linux/netlink.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <limits.h>
#include <wait.h>

#include "types.h"
#include "logger.h"
#include "config.h"
#include "vzerror.h"
#include "script.h"

#define NETLINK_UEVENT	31	/* from kernel/ve/vzevent.c */

static void child_handler(int signo)
{
	int pid, status;

	while ((pid = waitpid(-1, &status, WNOHANG)) > 0)
	{
		if (WIFEXITED(status))
			if (WEXITSTATUS(status) != 0)
				logger(-1, 0, "Child %d failed "
						"with exit code %d",
						pid, WEXITSTATUS(status));
			else
				logger(1, 0, "Child %d exited with success",
						pid);
		else if (WIFSIGNALED(status))
			logger(-1, 0, "Child %d killed by signal %d",
					pid, WTERMSIG(status));
	}
}

static int run_event_script(envid_t ctid, const char *event)
{
	char script[sizeof(LIB_SCRIPTS_DIR)*2];
	int pid;

	snprintf(script, sizeof(script), "%svzevent-%s",
			LIB_SCRIPTS_DIR, event);
	logger(1, 0, "Running %s event script", event);

	pid = fork();
	switch (pid) {
		case -1:
			logger(-1, errno, "Failed to fork()");
			return 1;
		case 0:
			exit(run_pre_script(ctid, script));
		default:
			logger(1, 0, "Forked child %d for %s event",
					pid, event);
	}
	return 0;
}

static int parse_event(char *buf)
{
	char *ctidp, *endp, *name;
	long int t;
	int len;
	envid_t ctid;
	const int min_event_len = 7; /* ve-stop */

	/* Parse CTID */
	if ((ctidp = strchr(buf, '@')) == NULL) {
		logger(-1, 0, "Bad message: can't find CTID");
		return -1;
	}

	/* Bail out of definitely bad/unknown events */
	len = ctidp - buf;
	if (len < min_event_len)
		goto ev_unknown;
	if (! (buf[0] == 'v') && (buf[1] == 'e') && (buf[2] == '-') )
		goto ev_unknown;
	name = buf + 3; /* Omit common "ve-" prefix */

	*ctidp++ = '\0';
	t = strtol(ctidp, &endp, 10);
	if (*endp != '\0') {
		logger(-1, 0, "Garbage in CTID in message: %s (endp=%s)",
				ctidp, endp);
		return -1;
	}
	if ((t <= 0) || (t > INT_MAX)) {
		logger(-1, 0, "Bad CTID in message: %s", ctidp);
		return -1;
	}
	ctid = (envid_t)t;
	set_log_ctid(ctid);
	logger(2, 0, "CTID = %d, event = %s (len=%d)", ctid, buf, len);

	/* Continue parsing event name */

	switch (len - 3) {
		case 4:
			if (strncmp(name, "stop", 4) == 0)
				goto ev_stop;
			goto ev_unknown;
		case 5:
			if (strncmp(name, "start", 5) == 0)
				goto ev_start;
			else if (strncmp(name, "mount", 5) == 0)
				goto ev_mount;
			goto ev_unknown;
		case 6:
			if (strncmp(name, "reboot", 6) == 0)
				goto ev_reboot;
			else if (strncmp(name, "umount", 6) == 0)
				goto ev_umount;
			goto ev_unknown;
		default:
			goto ev_unknown;
	}

ev_unknown:
	logger(-1, 0, "Unknown event: %s", buf);
	return -1;

ev_mount:
ev_umount:
ev_start:
	logger(2, 0, "Got %s event (ignored)", name);
	/* Do nothing */
	return 0;
ev_reboot:
	return run_event_script(ctid, "reboot");
ev_stop:
	return run_event_script(ctid, "stop");
}

static int read_events(int fd, struct sockaddr_nl *sa)
{
	int len;
	char buf[32];
	struct iovec iov = { buf, sizeof(buf) };
	struct msghdr msg;
	int ret;

	logger(0, 0, "Started");

	while (1) {
		memset(&msg, 0, sizeof(msg));
		msg.msg_name = (void *)&sa;
		msg.msg_namelen = sizeof(sa);
		msg.msg_iov = &iov;
		msg.msg_iovlen = 1;

		len = recvmsg(fd, &msg, 0);
		if (len > 0) {
			buf[len] = '\0';
			parse_event(buf);
			set_log_ctid(0);
		} else if (len < 0) {
			if (errno != EINTR) {
				logger(-1, errno, "Error in recvmsg() "
						" (ret=%d, errno=%d)",
						len, errno);
				ret = 1;
				break;
			}
		} else /* len==0 */ {
			logger(0, 0, "Connection closed");
			ret = 0;
			break;
		}
	}
	close(fd);
	logger(0, 0, "Exiting...");
	return ret;
}

static int daemon_read_events(int fd, struct sockaddr_nl *sa)
{
	if (daemon(0, 0) < 0) {
		logger(-1, errno, "Error in daemon()");
		return 1;
	}
	/* Now make logger stop printing to stdout/stderr */
	set_log_quiet(1);
	return read_events(fd, sa);
}

static int prepare_read_events(int daemonize)
{
	int fd;
	struct sockaddr_nl sa;
	struct sigaction act;

	fd = socket(AF_NETLINK, SOCK_RAW, NETLINK_UEVENT);
	if (fd < 0) {
		logger(-1, errno, "Error in socket()");
		return 1;
	}

	memset(&sa, 0, sizeof(sa));
	sa.nl_family = AF_NETLINK;
	sa.nl_groups = 1;
	if (bind(fd, (struct sockaddr *)&sa, sizeof(sa)) < 0) {
		if (errno == ENOENT)
			logger(-1, 0, "Looks like vzevent kernel module "
					"is not loaded; exiting.");
		else
			logger(-1, errno, "Error in bind()");
		close(fd);
		return 1;
	}

	sigemptyset(&act.sa_mask);
	act.sa_handler = child_handler;
	act.sa_flags = SA_NOCLDSTOP;
	sigaction(SIGCHLD, &act, NULL);

	if (daemonize != 0)
		return daemon_read_events(fd, &sa);
	else
		return read_events(fd, &sa);
}

static void usage()
{
	printf(
"Usage: vzeventd [options]\n"
"	-v	increase verbosity (can be used multiple times)\n"
"	-d	debug (do not daemonize, run in foreground)\n"
"	-h	print this help message\n"
	);
}

int main(int argc, char **argv)
{
	struct vps_param *param = NULL;
	int daemonize = 1;
	int opt, verbose = 0;

	while ((opt = getopt(argc, argv, "dv")) != -1) {
		switch (opt) {
		case 'd':
			daemonize = 0;
			verbose++;
			break;
		case 'v':
			verbose++;
			break;
		case 'h':
			usage();
			return 0;
		default:
			usage();
			return 1;
		}
	}

	param = init_vps_param();
	/* Read global config file */
	if (vps_parse_config(0, GLOBAL_CFG, param, NULL)) {
		fprintf(stderr, "Global configuration file %s not found\n",
				GLOBAL_CFG);
		return VZ_NOCONFIG;
	}

	init_log(param->log.log_file, 0, param->log.enable != NO,
			param->log.level + verbose,
			0, "vzeventd");

	return prepare_read_events(daemonize);
}
